appsTransfer/src/main/java/org/kinecosystem/appstransfer/presenter/SenderServiceBinder.kt

```diff
@@ -0,0 +1,153 @@
+                        mainThreadHandler.post {
+                            listener?.onBalanceReceived(balance)
+                        }
+                    }
+                } catch (balanceException: SendKinServiceBase.BalanceException) {
+                    mainThreadHandler.post {
+                        listener?.onBalanceFailed()
+                        //ignore if we dont get balance - user can send amount with no limit but will fail if it exceeds his balance
+                        //Log.d(TAG, "balanceException ${balanceException.message}")
+                    }
+                }
+            }
+        } else {
+            Log.e(TAG, "service must call bind first and wait until its bounded")
+        }
+    }
+
+    fun startSendKin(receiverAddress: String, senderAppName: String, amount: Int, memo: String, receiverPackage: String) {
+        if (isBounded) {
+            executorService.execute {
+                try {
+                    val kinTransferComplete: SendKinServiceBase.KinTransferComplete =
+                            transferService?.transferKin(receiverAddress, amount, memo)!!
+
+                    mainThreadHandler.post {
+                        // listener?.onTransferComplete()
+                    }
+
+                    try {
+//                        ReceiveKinNotifier.notifyTransactionCompleted(baseContext, receiverPackage,
+//                                kinTransferComplete.senderAddress, senderAppName, receiverAddress, amount,
+//                                kinTransferComplete.transactionId, kinTransferComplete.transactionMemo)
+                        Log.d(TAG, "Receiver was notified of transaction complete")
+                    } catch (e: ServiceConfigurationException) {
+                        Log.d(TAG, "Error notifying the receiver of transaction complete ${e.message}")
+                        e.printStackTrace()
+                    }
+                } catch (e: SendKinServiceBase.KinTransferException) {
+                    mainThreadHandler.post {
+                        // listener?.onTransferFailed()
+                    }
+                    Log.d(TAG, "Exception while transferring Kin,  SendKinServiceBase.KinTransferException ${e.message}")
+                    try {
+//                        ReceiveKinNotifier.notifyTransactionFailed(baseContext, receiverPackage,
+//                                e.toString(), e.senderAddress, senderAppName, receiverAddress, amount, memo)
+                        Log.d(TAG, "Receiver was notified of transaction failed")
+
+                    } catch (e: ServiceConfigurationException) {
+                        Log.d(TAG, "Error notifying the receiver of transaction failed ${e.message}")
+                        e.printStackTrace()
+                    }
+                }
+            }
+        } else {
+            Log.e(TAG, "service must call bind first and wait until its bounded")
+        }
+    }
+
+
+    fun bind() {
```

Comment:
you wanted to move logic from the activity ? this is why all is here?
Maybe the presenter should tell the activity to bind and so on, and then the activity will communicate all back to the presenter.

Reply:
Also shouldn't it throw `ServiceConfigurationException` in method header?

Reply:
added the throw
i remove it from the activity 
and it will be used by some modules so it like common part